### ISSUE TITLE
React Typescript integration

### ISSUE THREAD

**Issue by BDOXHN** [bumbeishvili/org-chart#182]
Hello! I really like your library, I would like to use it in the **React Typescript project.** I looked at an example for React Javascript and tried to adapt the code to typescript myself, having previously installed **@types/d3-org-chart**. Unfortunately, I didn't succeed. Do you have working code examples for React Typescript?
Thank you in advance!

**Comment by bumbeishvili** [comment]
Hi, no, I don't have a working example with React Typescript and nor intend typescript support in the future. 

Given the fact that methods are dynamically generated in this org chart I had difficulties adapting similar d3 charts with typescript and I tend to just disable typescript checks.

**Comment by marcellino-ornelas** [comment]
@BDOXHN what issues are you running into using this library with typescript? if you post some code I might be able to help. Im using this library with react and typescript 😃

**Comment by perdanafm** [comment]
Hai @marcellino-ornelas @bumbeishvili ,
I was trying to integrating the lib to typescript - nextjs project.
and i have this issue saying _**SyntaxError: Cannot use import statement outside a module**_

i tried to fix this by putting 'commonjs' on tsconfig but it doesn't work, anyone can help me with this issue ?



![image](https://user-images.githubusercontent.com/30098898/201732862-97430392-d622-49dd-976d-e1a5b07efa23.png)

**Comment by marcellino-ornelas** [comment]
This is our tsconfig.json

```json
{
	"include": ["src/**/*.ts", "src/**/*.tsx", "src/**/*.jsx", "src/**/*.js"],
	"exclude": ["node_modules/**"],
	"compilerOptions": {
		"target": "es6",
		"lib": ["es2016", "es2017", "es2019", "dom"],
		"module": "esnext",
		"moduleResolution": "node",
		"jsx": "react",
		"removeComments": false,
		"noImplicitAny": false,
		"resolveJsonModule": true,
		"baseUrl": ".",
		"paths": {
			"components/*": ["src/components/*"],
			"types/*": ["src/types/*"],
			"@redux/*": ["src/redux/*"],
			"@utils/*": ["src/utils/*"],
			"@testUtils/*": ["src/test/*"]
		},
		"skipLibCheck": true,
		"esModuleInterop": true
	}
}
```

and Im able to import the package like this

```tsx
// some-file.tsx
import { OrgChart } from 'd3-org-chart';
```

Not sure if nextJS has anything to do with this 🤔 Try to see if you can get it to work with something similar? if not want to post your tsconfig + how your importing the package

**Comment by marcellino-ornelas** [comment]
I also use `@types/d3-org-chart` not sure if that helps

https://www.npmjs.com/package/@types/d3-org-chart

**Comment by dlgoodchild** [comment]
I have a very basic React + Typescript implementation that may be useful to someone:
```
import React, {useLayoutEffect, useRef, useEffect} from 'react';
import ReactDOMServer from 'react-dom/server';
import {OrgChart} from 'd3-org-chart';
import { select } from "d3-selection";
import { HierarchyNode } from "d3-hierarchy";

const d3 = { select };

export interface Datum {

}

interface Props {
  data: Datum[];
  setClick: Function;
  onNodeClick: Function;
}

export default function MeterChartSample(props: Props) {
  const d3Container = useRef(null);
  const chart: OrgChart<Datum> = new OrgChart<Datum>();

  props.setClick((node: Datum) => chart!.addNode(node));

  useLayoutEffect(() => {
    if (props.data && d3Container.current) {
      chart
        .setActiveNodeCentered(false)
        .container(d3Container.current)
        .data(props.data)
        .onNodeClick((d) => {
          console.log(d, 'Id of clicked node ');
          props.onNodeClick(d);
        })
        .nodeContent((node, index, nodes, state): string => {
          return ReactDOMServer.renderToStaticMarkup(
            <div style={{width: '100%', height: '100%', border: '1px solid #666666', textAlign: 'center'}}>
              Sample ID {node.id}
            </div>
          )
        })
        .linkUpdate(function (d, i, arr) {
          // @ts-ignore
          d3.select(this)
            // @ts-ignore
            .attr("stroke", d => d.data._upToTheRootHighlighted ? '#152785' : 'black')
            // @ts-ignore
            .attr("stroke-width", d => d.data._upToTheRootHighlighted ? 6 : 2)

          // @ts-ignore
          if (d.data._upToTheRootHighlighted) {
            // @ts-ignore
            d3.select(this).raise()
          }
        })
        .layout("left")
        .compact(false)
        .nodeWidth((node: HierarchyNode<Datum>) => 150)
        .nodeHeight( (node: HierarchyNode<Datum>) => 60 )
        .render();
    }
  }, [props.data, d3Container.current]);

  return (
    <div>
      <div ref={d3Container} />
    </div>
  );
};
```

**Comment by Ziad-Mohamedd** [comment]
Hello @perdanafm 
I have the same issue as yours, did you find a solution?

The suggested solutions by @marcellino-ornelas and @dlgoodchild did not work for me, unfortunately.
Many thanks in advance.

**Comment by marcellino-ornelas** [comment]
@Ziad-Mohamedd want to make a reproducible example in https://stackblitz.com/ ? This way I can see the error and your configs so I can help you debug

**Comment by Ziad-Mohamedd** [comment]
@marcellino-ornelas https://stackblitz.com/edit/next-typescript-apze5m?file=pages/index.tsx 
:raised_hands:

**Comment by marcellino-ornelas** [comment]
@Ziad-Mohamedd I don't think this is a issue with this library but a issue with https://github.com/vercel/next.js/ server side rendering.

After taking a closer look I noticed this on another library

https://github.com/asyncapi/asyncapi-react/issues/177

which pointed me to
https://nextjs.org/docs/advanced-features/dynamic-import

I think whats going on is that next.js is trying to render this all server side but because this orgchart is written is es6 we would need to include it in some type of transpiling like webpack or typescript.

https://github.com/bumbeishvili/org-chart/blob/master/src/d3-org-chart.js

However after tinkering with the typescript configs I was still unable to get this package to get transpiled.

whats weird is that you can add something like this to the code and see the constructer get printed to the console

```
import { OrgChart } from 'node_modules/d3-org-chart/src/d3-org-chart.js';
```

![Screen Shot 2022-12-19 at 12 19 48 PM](https://user-images.githubusercontent.com/35247622/208513869-b8423fa9-ef61-4bb6-9941-2caecdb5b32e.png)

but it errors out because window is not defined because of the SSR. I would love to see what the next.js team has to say about this and there work around.

------
on a separate note

I happened to find this
https://stackoverflow.com/questions/66254521/how-can-i-import-browser-module-in-next-js-without-dynamic-import

and got the page to render again 🙌 
https://stackblitz.com/edit/next-typescript-nfppkg?file=package.json,pages%2Fchart.tsx,tsconfig.json,pages%2Findex.tsx

**Comment by marcellino-ornelas** [comment]
[perdanafm](https://github.com/perdanafm) ^^^^

**Comment by Ziad-Mohamedd** [comment]
@marcellino-ornelas 
First of all sorry for my late reply I was busy with work, as I am working on this as a side project,

Thank you so much for your work, It compiled with the `ssr: false`, good catch sir :ok_hand: 

Thanks again for your time. :raised_hands:

**Comment by EputraP** [comment]
Hi
can anyone help me, the code from @marcellino-ornelas works perfectly. 

Now I am confused on how to pass the data. 

I found this error in my code 
(Type '{ name: string; }' is not assignable to type 'Datum'.
  Object literal may only specify known properties, and 'name' does not exist in type 'Datum')
  
I used react TS

thankyou

**Comment by marcellino-ornelas** [comment]
@EputraP can you reproduce the issue in a Stackblitz or equivalent ? Happy to take a look
